src/object/srv_obj.c

```diff
@@ -1585,19 +1555,34 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
 
 	if (obj_rpc_is_fetch(rpc) && !spec_fetch &&
 	    daos_csummer_initialized(ioc->ioc_coc->sc_csummer)) {
-		rc = obj_fetch_csum_init(ioc->ioc_coc, orw, orwo);
+		/**
+		 * Allocate memory for the csum structures.
+		 * This memory and information will be used by VOS to put the checksums
+		 * in as it fetches the data's metadata from the btree/evtree.
+		 *
+		 * The memory will be freed in obj_rw_reply
+		 */
+		rc = daos_csummer_alloc_iods_csums(ioc->ioc_coc->sc_csummer, iods, iods_nr, false,
+						   NULL, &orwo->orw_iod_csums.ca_arrays);
+		if (rc >= 0) {
+			/*
+			 * daos_csummer_alloc_iods_csums returns the number of iod csums
+			 * allocated
+			 */
+			orwo->orw_iod_csums.ca_count = (uint64_t)rc;
+			rc                           = 0; /* reset the return code */
+		}
+
 		if (rc) {
-			D_ERROR(DF_UOID" fetch csum init failed: %d.\n",
-				DP_UOID(orw->orw_oid), rc);
+			D_ERROR(DF_UOID " unable to allocate iod csums: " DF_RC ".\n",
+				DP_UOID(orw->orw_oid), DP_RC(rc));
 			goto post;
 		}
 
 		if (ioc->ioc_coc->sc_props.dcp_csum_enabled) {
-			rc = csum_add2iods(ioh, orw->orw_iod_array.oia_iods,
-					   orw->orw_iod_array.oia_iod_nr,
-					   skips, ioc->ioc_coc->sc_csummer,
-					   orwo->orw_iod_csums.ca_arrays,
-					   orw->orw_oid, &orw->orw_dkey);
+			rc = csum_add2iods(ioh, iods, iods_nr, skips, ioc->ioc_coc->sc_csummer,
+					   orwo->orw_iod_csums.ca_arrays, orw->orw_oid,
+					   &orw->orw_dkey);
```

Comment:
ditto, the original code was try to use RPC's orw->orw_iod_array.oia_iods that include all original iods, but use the skip bitmap to skip some iods.